meta: run Windows tests on actual Windows VMs

We don't currently run the GitHub actions on Windows VMs. We currently just cross-compile things.

But GitHub doesn't charge extra for Windows (like they do for macOS), so just use Windows and then actually run the tests.

> But gotta fix all those failures before it can be merged:

I saw you started fixing some tests.

I also tried to debug net/netcheck.TestBasic. With this diff

```
diff --git a/net/netcheck/netcheck.go b/net/netcheck/netcheck.go
index 8433d1d..2e98924 100644
--- a/net/netcheck/netcheck.go
+++ b/net/netcheck/netcheck.go
@@ -467,8 +467,10 @@ func (c *Client) readPackets(ctx context.Context, pc net.PacketConn) {
        var buf [64 << 10]byte
        for {
                n, addr, err := pc.ReadFrom(buf[:])
+               fmt.Printf("READ: n=%v addr=%v err=%v\n", n, addr, err)
                if err != nil {
                        if ctx.Err() != nil {
+                               fmt.Printf("REA_: err=%v\n", ctx.Err())
                                return
                        }
                        c.logf("ReadFrom: %v", err)
@@ -1173,7 +1175,8 @@ func (rs *reportState) runProbe(ctx context.Context, dm *tailcfg.DERPMap, probe
 
        switch probe.proto {
        case probeIPv4:
-               rs.pc4.WriteTo(req, addr)
+               n, err := rs.pc4.WriteTo(req, addr)
+               fmt.Printf("WRITE: n=%v err=%v\n", n, err)
        case probeIPv6:
                rs.pc6.WriteTo(req, addr)
        default:
```

I can see this on windows

```
WRITE: n=0 err=write udp4 0.0.0.0:52655->127.0.0.1:52653: wsasendto: The requested address is not valid in its context.
WRITE: n=0 err=write udp4 0.0.0.0:52655->127.0.0.1:52653: wsasendto: The requested address is not valid in its context.
WRITE: n=0 err=write udp4 0.0.0.0:52655->127.0.0.1:52653: wsasendto: The requested address is not valid in its context.
READ: n=0 addr=<nil> err=read udp4 0.0.0.0:52655: use of closed network connection
REA_: err=context deadline exceeded
--- FAIL: TestBasic (1.00s)
    netcheck.go:185: probePortMapServices: me 192.168.56.1 -> gw 192.168.1.1
    netcheck.go:185: udp=false v4=false v6=false mapvarydest= hair= portmap=U derp=0
    netcheck_test.go:64: want UDP
    netcheck_test.go:67: expected 1 key in DERPLatency; got map[]
    netcheck_test.go:70: expected key 1 in DERPLatency; got map[]
    netcheck_test.go:73: expected GlobalV4 set
    netcheck_test.go:76: PreferredDERP = 0; want 1
    stuntest.go:63: STUN server shutdown
WRITE: n=0 err=write udp4 0.0.0.0:52659->127.0.0.1:52657: wsasendto: The requested address is not valid in its context.
WRITE: n=0 err=write udp4 0.0.0.0:52659->127.0.0.1:52657: wsasendto: The requested address is not valid in its context.
READ: n=0 addr=<nil> err=read udp4 0.0.0.0:52659: use of closed network connection
REA_: err=context deadline exceeded
FAIL
FAIL    tailscale.com/net/netcheck      1.309s
FAIL
```

and this on linux

```
=== RUN   TestBasic
WRITE: n=40 err=<nil>
READ: n=32 addr=127.0.0.1:44589 err=<nil>
READ: n=40 addr=127.0.0.1:37204 err=<nil>
    netcheck.go:185: probePortMapServices: me 192.168.1.27 -> gw 192.168.1.1
    netcheck.go:185: udp=true v6=false mapvarydest= hair=true portmap=U v4a=127.0.0.1:60230 derp=1 derpdist=1v4:0s
    stuntest.go:63: STUN server shutdown
--- PASS: TestBasic (0.10s)
PASS
ok      tailscale.com/net/netcheck      0.106s
```

Feel free to fix it, if you know what the problem is. But I will continue my debug, if I don't hear from you.

Alex

That's also a netns issue, very similar to the one I fixed in the last day,
but for UDP where the address isn't known until after the netns dial.


On Thu, Oct 15, 2020 at 1:39 AM Alex Brainman <notifications@github.com>
wrote:

> But gotta fix all those failures before it can be merged:
>
> I saw you started fixing some tests.
>
> I also tried to debug net/netcheck.TestBasic. With this diff
>
> diff --git a/net/netcheck/netcheck.go b/net/netcheck/netcheck.go
> index 8433d1d..2e98924 100644
> --- a/net/netcheck/netcheck.go
> +++ b/net/netcheck/netcheck.go
> @@ -467,8 +467,10 @@ func (c *Client) readPackets(ctx context.Context, pc net.PacketConn) {
>         var buf [64 << 10]byte
>         for {
>                 n, addr, err := pc.ReadFrom(buf[:])
> +               fmt.Printf("READ: n=%v addr=%v err=%v\n", n, addr, err)
>                 if err != nil {
>                         if ctx.Err() != nil {
> +                               fmt.Printf("REA_: err=%v\n", ctx.Err())
>                                 return
>                         }
>                         c.logf("ReadFrom: %v", err)
> @@ -1173,7 +1175,8 @@ func (rs *reportState) runProbe(ctx context.Context, dm *tailcfg.DERPMap, probe
>
>         switch probe.proto {
>         case probeIPv4:
> -               rs.pc4.WriteTo(req, addr)
> +               n, err := rs.pc4.WriteTo(req, addr)
> +               fmt.Printf("WRITE: n=%v err=%v\n", n, err)
>         case probeIPv6:
>                 rs.pc6.WriteTo(req, addr)
>         default:
>
> I can see this on windows
>
> WRITE: n=0 err=write udp4 0.0.0.0:52655->127.0.0.1:52653: wsasendto: The requested address is not valid in its context.
> WRITE: n=0 err=write udp4 0.0.0.0:52655->127.0.0.1:52653: wsasendto: The requested address is not valid in its context.
> WRITE: n=0 err=write udp4 0.0.0.0:52655->127.0.0.1:52653: wsasendto: The requested address is not valid in its context.
> READ: n=0 addr=<nil> err=read udp4 0.0.0.0:52655: use of closed network connection
> REA_: err=context deadline exceeded
> --- FAIL: TestBasic (1.00s)
>     netcheck.go:185: probePortMapServices: me 192.168.56.1 -> gw 192.168.1.1
>     netcheck.go:185: udp=false v4=false v6=false mapvarydest= hair= portmap=U derp=0
>     netcheck_test.go:64: want UDP
>     netcheck_test.go:67: expected 1 key in DERPLatency; got map[]
>     netcheck_test.go:70: expected key 1 in DERPLatency; got map[]
>     netcheck_test.go:73: expected GlobalV4 set
>     netcheck_test.go:76: PreferredDERP = 0; want 1
>     stuntest.go:63: STUN server shutdown
> WRITE: n=0 err=write udp4 0.0.0.0:52659->127.0.0.1:52657: wsasendto: The requested address is not valid in its context.
> WRITE: n=0 err=write udp4 0.0.0.0:52659->127.0.0.1:52657: wsasendto: The requested address is not valid in its context.
> READ: n=0 addr=<nil> err=read udp4 0.0.0.0:52659: use of closed network connection
> REA_: err=context deadline exceeded
> FAIL
> FAIL    tailscale.com/net/netcheck      1.309s
> FAIL
>
> and this on linux
>
> === RUN   TestBasic
> WRITE: n=40 err=<nil>
> READ: n=32 addr=127.0.0.1:44589 err=<nil>
> READ: n=40 addr=127.0.0.1:37204 err=<nil>
>     netcheck.go:185: probePortMapServices: me 192.168.1.27 -> gw 192.168.1.1
>     netcheck.go:185: udp=true v6=false mapvarydest= hair=true portmap=U v4a=127.0.0.1:60230 derp=1 derpdist=1v4:0s
>     stuntest.go:63: STUN server shutdown
> --- PASS: TestBasic (0.10s)
> PASS
> ok      tailscale.com/net/netcheck      0.106s
>
> Feel free to fix it, if you know what the problem is. But I will continue
> my debug, if I don't hear from you.
>
> Alex
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tailscale/tailscale/issues/50#issuecomment-708995563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAUPMTEBSKUUPDFJU35QDSK2YNVANCNFSM4KUZIYOQ>
> .
>

The only remaining broken test is tailscale.com/portlist

```
--- FAIL: TestGetList (0.03s)
    portlist_test.go:20: addProcesses: netstat: exit status 1 ("")
--- FAIL: TestIgnoreLocallyBoundPorts (0.03s)
    portlist_test.go:41: addProcesses: netstat: exit status 1 ("")
FAIL
FAIL    tailscale.com/portlist  0.091s
```

The problem with the test is that it uses `netstat -nab` command, that can only run as Administrator. The test PASSes, if I run it from Administrator command window. What should I do? I don't see how these tests will run as Administrator ever. So the only thing that I can think of is to just skip the test.

I also don't understand how this code is useful, if it needs to be run by admin. I take it is running as a Windows service.

Alex

--- a/.github/workflows/windows.yml
+++ b/.github/workflows/windows.yml
@@ -1,0 +1,44 @@
+name: Windows
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - '*'
+
+jobs:
+  test:
+    runs-on: windows-latest
+
+    if: "!contains(github.event.head_commit.message, '[ci skip]')"
+
+    steps:
+
+    - name: Install Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.15.x
+
+    - name: Checkout code
+      uses: actions/checkout@v2
+
+    - name: Test
+      run: go test ./...
+
+    - uses: k0kubun/action-slack@v2.0.0
+      with:
+        payload: |
+          {
+            "attachments": [{
+              "text": "${{ job.status }}: ${{ github.workflow }} <https://github.com/${{ github.repository }}/commit/${{ github.sha }}/checks|${{ env.COMMIT_DATE }} #${{ env.COMMIT_NUMBER_OF_DAY }}> " +
+                      "(<https://github.com/${{ github.repository }}/commit/${{ github.sha }}|" + "${{ github.sha }}".substring(0, 10) + ">) " +
+                      "of ${{ github.repository }}@" + "${{ github.ref }}".split('/').reverse()[0] + " by ${{ github.event.head_commit.committer.name }}",
+              "color": "danger"
+            }]
+          }
+      env:
+        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
+      if: failure() && github.event_name == 'push'
+

